core/unit_test/TestGraph.hpp

```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+  static void compute(const Kokkos::Cuda& exec, DstType* const dst,
+                      const Srcs* const... srcs) {
+    set_to<GraphNodeType::CAPTURE>
+        <<<dim3(1, 1, 1), dim3(1, 1, 1), 0, exec.cuda_stream()>>>(dst, srcs...);
+  }
+
+  EXTERNAL_CAPTURE_ADD(cuda_capture, Cuda)
+};
+#endif
+#if defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)
+template <>
+struct ExternalCapture<Kokkos::HIP> {
+  template <typename DstType, typename... Srcs>
+  static void compute(const Kokkos::HIP& exec, DstType* const dst,
+                      const Srcs* const... srcs) {
+    set_to<GraphNodeType::CAPTURE>
+        <<<dim3(1, 1, 1), dim3(1, 1, 1), 0, exec.hip_stream()>>>(dst, srcs...);
+  }
+
+  EXTERNAL_CAPTURE_ADD(hip_capture, HIP)
+};
+#endif
+#if defined(KOKKOS_ENABLE_SYCL) && defined(SYCL_EXT_ONEAPI_GRAPH)
+template <>
+struct ExternalCapture<Kokkos::SYCL> {
+  template <typename DstType, typename... Srcs>
+  static void compute(const Kokkos::SYCL& exec, DstType* const dst,
+                      const Srcs* const... srcs) {
+    exec.sycl_queue().submit([&](sycl::handler& cgh) {
+      cgh.parallel_for(sycl::range<1>(1), [=](int) {
+        dst[0] +=
+            (srcs[0] + ...) + static_cast<DstType>(GraphNodeType::CAPTURE);
+      });
+    });
+  }
+
+  EXTERNAL_CAPTURE_ADD(sycl_capture, SYCL)
+};
+#endif
+
+template <typename Exec>
+std::enable_if_t<GraphNodeTypes<Exec>::support_capture == false, void>
+test_graph_capture() {
+  GTEST_SKIP() << "The execution space " << Kokkos::Impl::TypeInfo<Exec>::name()
+               << " does not support capture.";
+}
+
+template <typename Exec>
+std::enable_if_t<GraphNodeTypes<Exec>::support_capture == true, void>
+test_graph_capture() {
+  using view_r1_t = Kokkos::View<int[5], Exec>;
+  using view_r0_t = Kokkos::View<int, Exec>;
+
+  const auto execs = Kokkos::Experimental::partition_space(Exec{}, 1, 1, 1, 1);
+
+  const auto& exec       = execs.at(0);
+  const auto& exec_graph = execs.at(1);
+  const auto& exec_left  = execs.at(2);
+  const auto& exec_right = execs.at(3);
+
+  constexpr int offset_left = 123, offset_right = 456;
+
+  const view_r1_t data(
+      Kokkos::view_alloc(exec, "data used in the captured kernel"));
+  exec.fence("Wait for data to be initialized.");
+
+  // clang-format off
+#define TAKE_R0_AT(index) const view_r0_t data_##index(Kokkos::subview(data, index));
+
+  TAKE_R0_AT(0) TAKE_R0_AT(1) TAKE_R0_AT(2) TAKE_R0_AT(3) TAKE_R0_AT(4)
+
+  auto graph = Kokkos::Experimental::create_graph<Exec>(exec_graph);
+  auto root  = Kokkos::Impl::GraphAccess::create_root_ref(graph);
+  // clang-format on
+
+  auto memset_left = root.then_parallel_for(
+      Kokkos::RangePolicy<Exec>(exec_left, 0, 1),
+      SetViewToValueFunctor<Exec, int>{data_0, offset_left});
+
+  auto memset_right = root.then_parallel_for(
+      Kokkos::RangePolicy<Exec>(exec_right, 0, 1),
+      SetViewToValueFunctor<Exec, int>{data_4, offset_right});
```

Comment:
Is it right to say that because we're running this test on device backends only, the execution space instance that we pass here ultimately is not used? If that's right, should we remove the `exec_left` and `exec_right`? And could we have e.g. an `exec` to submit the graph on, and and `exec_capture` that we use for the capture?

Reply:
@romintomasetti and I discussed about it. On a backend with a native graph implementation, the current impact of the `exec` passed to `then_parallel_for` is:
- With the constant and global launch mechanisms, the passed `exec` is used to copy the driver to device. With the local launch mechanism, there is no such use.
- In particular on `Cuda`, multi GPU graphs are supported. For such use cases, the passed `exec` can determine the particular device on which the graph node will be executed.   

We decided it makes sense to keep the `exec_left` and `exec_right` arguments because it's a general way of coding that is agnostic of which launch mechanism will be used. And it anticipates multi GPU use cases.